sycl/doc/extensions/SPIRV/SPV_INTEL_hw_thread_queries.asciidoc

```diff
@@ -0,0 +1,174 @@
+- Mateusz Belicki, Intel +
+
+== Notice
+
+Copyright (c) 2021 Intel Corporation.  All rights reserved.
+
+== Status
+
+Working Draft
+
+This is a preview extension specification, intended to provide early access to a
+feature for review and community feedback. When the feature matures, this
+specification may be released as a formal extension.
+
+
+Because the interfaces defined by this specification are not final and are
+subject to change they are not intended to be used by shipping software
+products. If you are interested in using this feature in your software product,
+please let us know!
+
+== Version
+
+[width="40%",cols="25,25"]
+|========================================
+| Last Modified Date | {docdate}
+| Revision           | A
+|========================================
+
+== Dependencies
+
+This extension is written against the SPIR-V Specification,
+Version 1.5 Revision 5.
+
+This extension requires SPIR-V 1.0.
+
+== Overview
+
+The extension adds BuiltIn Decorations to query maximum number of H/W thread IDs per sub-device, sub-device ID and H/W thread ID within the sub-device. These instructions combined can be used to
+calculate unique H/W thread ID which is executing the code.
+
+
+== Extension Name
+
+
+To use this extension within a SPIR-V module, the appropriate *OpExtension* must
+be present in the module:
+
+[subs="attributes"]
+----
+OpExtension "{extension_name}"
+----
+
+== New Capabilities
+
+This extension introduces new capabilities:
+
+[subs="attributes"]
+----
+{capability_name}
+----
+
+== New BuiltIn Decorations
+
+BuiltIn Decorations added under the *{capability_name}* capability:
+
+----
+
+SubDeviceIDINTEL
+HWThreadIDINTEL
+MaxHWThreadIDPerSubDeviceINTEL
+
+----
+
+== Token Number Assignments
+
+[width="40%"]
+[cols="70%,30%"]
+[grid="rows"]
+|====
+|*{capability_name}* | {capability_token}
+|*SubDeviceIDINTEL*  | {SubDeviceIDINTEL_token}
+|*HWThreadIDINTEL*  | {HWThreadIDINTEL_token}
+|*MaxHWThreadIDPerSubDeviceINTEL*  | {MaxHWThreadIDPerSubDeviceINTEL_token}
+|====
+
+== Modifications to the SPIR-V Specification, Version 1.5
+
+
+=== Capabilities
+
+Modify Section 3.31, Capability, adding rows to the Capability table:
+
+--
+[options="header"]
+|====
+2+^| Capability ^| Implicitly Declares 
+| {capability_token} | *{capability_name}*
+| Reserved. +
+ +
+See also extension: *{extension_name}*
+|====
+--
+
+=== Decorations
+
+Modify Section 3.20, Decoration, adding <<Literal>> _BuiltIn parameter_ to <<Extra Operands>> of *BuiltIn* decoration.
+
+Modify Section 3.21, BuiltIn, adding these rows to the BuiltIn table:
+
+==== 3.21 BuiltIn
+
+--
+[options="header"]
+|====
+2+^| BuiltIn 2+^| Enabling Capabilities
+| {SubDeviceIDINTEL_token} | [[SubDeviceIDINTEL]]*SubDeviceIDINTEL* +
+ ID of the sub-device on which the code running in *Kernel* +
+<<Execution Model,Execution Model>>. The BuiltIn decoration takes additional _Literal_ +
+ *Affinity* parameter, see the client API specification for more detail. +
```

Comment:
I think that we have to describe the possible values of the affinity parameter somewhere. My understanding is that we will support two affinity modes: (1) ignoring current runtime affinity setting and returning the physical ID value, (2) respecting the runtime settings and returning logical ID. Those modes will be represented by literal integer value, that will be mapped to some enumeration. I believe we have to specify this enumeration and agree which value means what. Here's similar mapping to define matrix layout in joint matrix spec: https://gitlab.devtools.intel.com/OpenCL/opencl-extension-drafts/-/blob/76c42e92d3f223a6b5b2fee03a63030317c73941/SPV_INTEL_joint_matrix.asciidoc#user-content-matrix-layout